### ISSUE TITLE
[Bug] Multiple instances

### ISSUE THREAD

**Issue by 1368129224** [iyear/tdl#564]
### Describe the bug

Run the tdl command in two terminals at the same time, one of them returns an error
`Current database is used by another process, please terminate it first`

### To Reproduce

1. run `tdl dl  -d /mnt/xxx/  -u https://t.me/xxxxxx` in terminal 1
2. run `tdl dl  -d /mnt/xxx/  -u https://t.me/yyyyyy` in terminal 2 at the same time

### Expected behavior

Downloading files in two terminals at the same time

### Version

Version: 0.16.1
Commit: bb2c69b
Date: 2024-03-01T13:21:31Z

go1.21.7 linux/amd64

### Which OS are you running tdl on?

Linux

**Comment by 1368129224** [comment]
Noted that a similar issue already exists, but it is still not possible to use multiple instances at the same time.
https://github.com/iyear/tdl/issues/424

**Comment by bolavefasfas** [comment]
> ### Describe the bug
> Run the tdl command in two terminals at the same time, one of them returns an error `Current database is used by another process, please terminate it first`
> 
> ### To Reproduce
> 1. run `tdl dl  -d /mnt/xxx/  -u https://t.me/xxxxxx` in terminal 1
> 2. run `tdl dl  -d /mnt/xxx/  -u https://t.me/yyyyyy` in terminal 2 at the same time
> 
> ### Expected behavior
> Downloading files in two terminals at the same time
> 
> ### Version
> Version: 0.16.1 Commit: [bb2c69b](https://github.com/iyear/tdl/commit/bb2c69bce24d431a6d7ca4241e115b8751a89080) Date: 2024-03-01T13:21:31Z
> 
> go1.21.7 linux/amd64
> 
> ### Which OS are you running tdl on?
> Linux
> 
> ### Additional context
> _No response_

Use a ubuntu docker container and install tdl in that and login account again so that you get 2 or more isolated logins 
Thats what I have been doing 😂

If you want I can share my setup

Or spin up vm/lxc

**Comment by XMLHexagram** [comment]
In my view, this is a feature not a bug. 

Author define and deal with this expected error in `main.go`
https://github.com/iyear/tdl/blob/0174e6dbed190606ea4ae3b0c79b2ca96169cc49/main.go#L20-L23

----------------------------------------------

Go deeper, `tdl` use [bblot](https://github.com/etcd-io/bbolt) behind to save and provide necessary data.

In `bblot`'s doc, it says
> Please note that Bolt obtains a file lock on the data file so multiple processes cannot open the same database at the same time.

In `bblot`'s doc we can know that only `bblot` only support one writable access at the same time.

So in conclusion, it is almost impossible to enable two or more download at the same time in the same namespace.

**Comment by 1368129224** [comment]
> > ### Describe the bug
> > Run the tdl command in two terminals at the same time, one of them returns an error `Current database is used by another process, please terminate it first`
> > ### To Reproduce
> > 
> > 1. run `tdl dl  -d /mnt/xxx/  -u https://t.me/xxxxxx` in terminal 1
> > 2. run `tdl dl  -d /mnt/xxx/  -u https://t.me/yyyyyy` in terminal 2 at the same time
> > 
> > ### Expected behavior
> > Downloading files in two terminals at the same time
> > ### Version
> > Version: 0.16.1 Commit: [bb2c69b](https://github.com/iyear/tdl/commit/bb2c69bce24d431a6d7ca4241e115b8751a89080) Date: 2024-03-01T13:21:31Z
> > go1.21.7 linux/amd64
> > ### Which OS are you running tdl on?
> > Linux
> > ### Additional context
> > _No response_
> 
> Use a ubuntu docker container and install tdl in that and login account again so that you get 2 or more isolated logins Thats what I have been doing 😂
> 
> If you want I can share my setup
> 
> Or spin up vm/lxc

Is it possible to use the same account in multiple `tdl` and download file at the same time this way?

**Comment by 1368129224** [comment]
> In my view, this is a feature not a bug.
> 
> Author define and deal with this expected error in `main.go`
> 
> https://github.com/iyear/tdl/blob/0174e6dbed190606ea4ae3b0c79b2ca96169cc49/main.go#L20-L23
> 
> Go deeper, `tdl` use [bblot](https://github.com/etcd-io/bbolt) behind to save and provide necessary data.
> 
> In `bblot`'s doc, it says
> 
> > Please note that Bolt obtains a file lock on the data file so multiple processes cannot open the same database at the same time.
> 
> In `bblot`'s doc we can know that only `bblot` only support one writable access at the same time.
> 
> So in conclusion, it is almost impossible to enable two or more download at the same time in the same namespace.

Thanks for the explanation from a technical point of view.
I found a similar issue https://github.com/iyear/tdl/issues/424 where the author said it was fixed in v0.14.0, but it looks like it's still not possible to use multiple instances at the same time because of `bblot`?

**Comment by XMLHexagram** [comment]
@1368129224 

First question:
> I found a similar issue https://github.com/iyear/tdl/issues/424 where the author said it was fixed in v0.14.0, but it looks like it's still not possible to use multiple instances at the same time because of bblot?

In v0.14.0 release note we can see:
> 296d928: feat(main): humanize some errors (@iyear)
> https://github.com/iyear/tdl/commit/684341834b57579ce422ca0ad738891dfef86f90: feat(storage): switch default storage to bolt (@iyear)

So I believe, the **fix** in #424 means switch to `bblot` and humanize the multiply instance error.

--------------------------------

Second question:

> Is it possible to use the same account in multiple tdl and download file at the same time this way?

Of course you can. There is no limit on it.
